k2/csrc/array_ops_benchmark.cu

```diff
@@ -0,0 +1,36 @@
+/**
+ * @brief Benchmarks for k2 APIs.
+ *
+ * @copyright
+ * Copyright (c)  2020  Xiaomi Corporation (authors: Fangjun Kuang)
+ *
+ * @copyright
+ * See LICENSE for clarification regarding multiple authors
+ */
+
+#include "benchmark/benchmark.h"
+#include "k2/csrc/array_ops.h"
+#include "k2/csrc/timer.h"
+
+template <typename T>
+static void BM_ExclusiveSum(benchmark::State &state) {
+  using namespace k2;  // TODO(fangjun): get rid of `using`  // NOLINT
+  ContextPtr context = GetCudaContext();
+  Timer timer(context);
+  int32_t num_bytes_processed = 0;
+  for (auto _ : state) {
+    int32_t dim = state.range(0);
+    Array1<T> src = RandUniformArray1<T>(context, dim, -1000, 1000);
+    timer.Reset();
+    Array1<T> dst = ExclusiveSum(src);
+    state.SetIterationTime(timer.Elapsed());  // in seconds
+    num_bytes_processed += dim * sizeof(T);
+  }
+  state.SetBytesProcessed(num_bytes_processed);
+}
+
+BENCHMARK_TEMPLATE(BM_ExclusiveSum, int32_t)
+    ->Range(8, 8 << 10)
```

Comment:
Compared to use those `marco` (or `function`?) from google benchmark, I prefer to declare problem sizes manually here, so the reader can easily know what problem sizes we'll run.

(I have to check the documentation of google benchmark to know what this line means: `By default the arguments in the range are generated in multiples of eight and the command above selects [ 8, 64, 512, 4k, 8k ]`)

Reply:
You can find the problem sizes at the output:
```
BM_ExclusiveSum<int32_t>/8/manual_time 
BM_ExclusiveSum<int32_t>/64/manual_time 
BM_ExclusiveSum<int32_t>/512/manual_time 
BM_ExclusiveSum<int32_t>/4096/manual_time 
BM_ExclusiveSum<int32_t>/8192/manual_time
```

`8->64->512->4096->8192`.

Reply:
of course we can as you have pasted the output here. But what I mean is: reader should be able to figure out the sizes by just reading code instead of running code...

Reply:
`Range` is to free us specifying the sizes exhaustively. What do you think are the sutiable sizes that should be written down manually?

Reply:
I'd like to add `100, 500,1000, 2000, 5000, 10000` etc here, maybe +m whre m is a small random number which can be negative or positive. We need to run multiple times for a particular size and get the average time (e.g. with the random number `m`, for `size=1000`, we run 5 times, we get the result of size = 999, size == 1002, size = 1001, size= 997, size = 1001, and we compute the average time)

Another point here is: we should not just run benchmark on those 2^n sizes (that is you did here with `Range`), they may be faster than real normal case (non 2^n sizes). 

 @danpovey as he may have some more insights.

Reply:
BTW, for kernels performace tests, we usually add a `warm-up` before the real benchmark starting in case the first kernel lanuch takes longer time. e.g. 
```
mykenerl(); // just a warm-up
Time t;
for(int32_t i = 0; i != 10; ++i){
  mykenerl();
}
t.Elapsed()
```

But I'm not sure if we need to do this for our op testings. Maybe @danpovey can give some comments here.

Reply:
>  But what I mean is: reader should be able to figure out the sizes by just reading code instead of running code

Users that have used `google/benchmark` before should be able to figure out the sizes from `Range`.
For users who have never used `google/bechmark`, I recommend them to refer to the documentation
and code of `google/benchmark`.

Reply:
mm, you're just trying to ask our code readers learn more nonessential tools....Anyway, it's up to you, but I believe those 2^n sizes you set are definitely not right for a typical benchmark running (especially for cuda kenerls)

Reply:
@csukuangfj I don't like to assume people read the documentation of projects we require because:
  - in most cases they won't have
  - we're probably not providing a pointer
  - there may be a lot of documentation and it's too much reading to be able to figure out what's needed for our stuff.
I'd rather provide some brief comments in our code to help people understand what's going on.

Reply:
> I'd like to add 100, 500,1000, 2000, 5000, 10000 etc here, maybe +m whre m is a small random number which can be negative or positive

If random numbers come  into scope, will it be difficult to compare benchmark results in different settings, i.e.,
release vs debug?

> We need to run multiple times for a particular size and get the average time (e.g. with the random number m, for size=1000, we run 5 times, we get the result of size = 999, size == 1002, size = 1001, size= 997, size = 1001, and we compute the average time

You can see from the output posted above
```
----------------------------------------------------------------------------------------------------
Benchmark                                          Time             CPU   Iterations
----------------------------------------------------------------------------------------------------
BM_ExclusiveSum<int32_t>/8/manual_time         15451 ns        38989 ns        43121
```

It is the average of `43121` iterations. The number of iterations needed is computed automatically
by google/benchmark.


> BTW, for kernels performace tests, we usually add a warm-up before the real benchmark starting in case the first kernel lanuch takes longer time. e.g.

The results are the average of thousands of iterations. Anyway, warm up is added as you suggested.

Reply:
>  but I believe those 2^n sizes you set are definitely not right for a typical benchmark running (especially for cuda kenerls)

Can you explain why the sizes of power of two are.NOT.right for benchmarks? I think it can show how the algorithm
scales with the increasing number of items in logarithmic scale.

Reply:
> I don't like to assume people read the documentation of projects we require because:

I agree, but the point is the relevant code is about implementation details. If the user/developer is
really interested in it, he/she can read the API documentation.

We are using `cub::DeviceScan::ExclusiveSum` in k2, but I think we only add documentation
to `k2::ExclusiveSum`. If someone is interested in `cub::DeviceScan::ExclusiveSum`, he/she
should try to find its doc provided by cub.

Reply:
> > but I believe those 2^n sizes you set are definitely not right for a typical benchmark running (especially for cuda kenerls)
> 
> Can you explain why the sizes of power of two are.NOT.right for benchmarks? I think it can show how the algorithm
> scales with the increasing number of items in logarithmic scale.

Well, Cuda does computation with `warps`  instead of a single thread, normally a warp of thread is 16 or 32 (depends on CUDA ARCH, IIRC). If we are using 2^n sizes, actually, there may be no idle thread, this is not real for most normal cases. Besides, we always to start block with size of 2^m (e.g. 128, 256) to leverage cuda arch (IIRC, apply for such block will be faster than non 2^n sizes, according to CUDA arch), testing with 2^n problems mean we may get the highest performace, this is not reasonal for a typical benchmark running.

Reply:
I see, thanks.

Then the problem was that I used 8 and its default multiplier `8` in `Range(8, 8 << 10)` .
`Range` would be fine here if we change its default multiplier.

Reply:
BTW, IIRC ,block size should be always a round multiple of the warp size (e.g. 32); otherwise, we are wasting some computation resources as cuda always make it to be a multiple of 32.

Reply:
> I see, thanks.
> 
> Then the problem was that I used 8 and its default multiplier `8` in `Range(8, 8 << 10)` .
> `Range` would be fine here if we change its default multiplier.

mm, of course we can, but I still prefer to add some random numbers here, the disadvantage of `Range` stuff here is: no matter what `multiplier` we use, we are still testing with a particular distribution of data (thinking about `size % 32==warp_size`). But as you said, it's hard to do this in google benchmark. so I'm not sure about this (if google benchmark cannot, should we still use it for benchmark or not? or we just don't use some fixed problem sizes) @danpovey any suggestions?